### ISSUE TITLE
Implementing open standards and protocols for server-to-server federation

### ISSUE THREAD

**Issue by strypey** [Qbix/Platform-History-v2#768]
Found this project through the [link EGreg posted on Hacker News](https://www.youtube.com/watch?v=pZ1O_gmPneI&app=desktop). QBix seems like a [federated social network](http://fediverse.gitlab.io/) project, which is awesome. Have you implemented any of the [open standards and protocols for server-to-server federation](https://medium.com/we-distribute/a-quick-guide-to-the-free-network-c069309f334?source=collection_home---6------3----------------)? If not, are you planning to?

**Comment by EGreg** [comment]
Yes, we are eventually planning to integrate support for various protocols including:

- Scuttlebutt
- Matrix.org
- Mastodon

**Comment by strypey** [comment]
That's awesome. I presume by Mastodon you mean ActivityPub, as that's the main standard it uses for server>server communication. It also remains backwards compatible with OStatus, which AP was developed to replace, but only a handful of other projects ever implemented OS and all of them have or are moving to support AP (**EDIT**: *as of 2019-08-02 Mastodon is deprecating OStatus support, even though a couple of the legacy OStatus apps haven't implemented AP yet*).

Scuttlebutt/ SSB is a cool tech and community, but at the risk of nitpicking, it's a client>client protocol (P2P/ distributed) not server>server (federated). Are you intending to have QBix servers operate as SSB "pubs", or are the QBix clients that you envision connecting directly to SSB clients?

**Comment by EGreg** [comment]
This is where it gets a bit hairy. See, Qbix servers will be able to host accounts for one person, or many people. Most accounts would, however, be end-to-end encrypted, so the keys would only be held on certain computers. The "Hubs" would not have access to encrypted information. They would only be used for leaving messages in various mailboxes (steganography), where others would pick them up and decrypt them.

We are still figuring out how this would interoperate with the open protocols, whether we could support the full functionality. It's likely we would have to launch a protocol forked from SSB.

**Comment by strypey** [comment]
This sounds *sort of* similar to the way the [SSB protocol](https://www.scuttlebutt.nz/protocol) works (as I understand it so far).

> Qbix servers will be able to host accounts for one person, or many people.

In Scuttlebutt, accounts are hosted on the users' devices. The 'pubs' are supplementary ways of discovering people to connect with and syncing messages with them when their devices are offline. Otherwise it's a serverless/ P2P protocol. Pubs work sort of like keyservers for PGP public keys; users can upload their SSB public key to any pub they like, or none.

> Most accounts would, however, be end-to-end encrypted, so the keys would only be held on certain computers. The "Hubs" would not have access to encrypted information.

In Scuttlebutt, *all* accounts are E2EE, and the keys are only held on the devices hosting the accounts.

>  They would only be used for leaving messages in various mailboxes (steganography), where others would pick them up and decrypt them.

This sounds similar to how SSB works, except that as mentioned above, users can leave messages in each others' mailboxes on any pub they are connected with, or just exchange them directly with each other P2P, as with Briar, Tox, or Jami.

**Comment by aschrijver** [comment]
It would be _**fantastic**_ if you would implement ActivityPub support in one form or another and become part of the Fediverse with its (currently) 4 million users. Like @strypey I help maintain the research wiki for [fediverse.party](https://fediverse.party), which tracks AP apps in various stages of development. Would love :heart: to add you to [this watchlist](https://git.feneas.org/feneas/fediverse/-/wikis/watchlist-for-activitypub-apps).

**Comment by EGreg** [comment]
@aschrijver - I just ran across this. Sounds exciting, and we have it on our roadmap now! Is there an updated link to the watchlist? I hear Mastodon had a huge uptick in users since Elon bought Twitter.

PS: How do you see interoperation between ActivityPub and Matrix Protocol?

**Comment by aschrijver** [comment]
Since Feneas has shut down, the 3 Fediverse related lists have moved to [Delightful Club](https://delightful.club).
Also have a look at [Kazarma](https://gitlab.com/kazarma/kazarma/) for an AP/Matrix bridge project, and certainly join [SocialHub](https://socialhub.activitypub.rocks) developer community :)

(Btw, on a high-end laptop the QBix website brought my Firefox to a crawl with all the interactivity, almost crashing)

**Comment by EGreg** [comment]
Which URL? Just visiting it?

**Comment by aschrijver** [comment]
Yes, just visiting the landing page.

**Comment by EGreg** [comment]
@aschrijver is the qbix.com website still bringing your laptop to a crawl?

**Comment by aschrijver** [comment]
No, everything fine now. Only the sparkles on the mouse pointer are distracting when reading the site.

(Btw, sorry I didn't notice before that QBix is a blockchain project, as these aren't currently candidates for the delightful project)

**Comment by strypey** [comment]
@EGreg
>How do you see interoperation between ActivityPub and Matrix Protocol?

There are a few experimental projects trying to bridge the two networks, including:
https://gitlab.com/kazarma/kazarma/

One challenge is that existing usage of the two protocols is quote different. Matrix is primarily used for group rooms and private messaging. ActivityPub is primarily used for public-facing social media, mainly short text posts with occasional image attachments a la Mastodon.
